### ISSUE TITLE
Connection to device is systematically done twice

### ISSUE THREAD

**Issue by chanqueo** [square/pylink#113]
When calling connect() function, connection to device is done twice.
Example with the following code:

```
self.jlink = pylink.JLink()
ip_addr = socket.gethostbyname(hostname)
self.jlink.open(ip_addr=f"{ip_addr}:19020")
self.jlink.set_tif(pylink.JLinkInterfaces.SWD)
self.jlink.connect(chip_name="EFR32MG12PXXXF1024", speed="auto", verbose=True)
```

Call to function connect() will give following logs (notice verbose is set to True)
```
13:38:13.257;INFO;pylink.jlink;Device "EFR32MG12PXXXF1024" selected.
13:38:13.359;INFO;pylink.jlink;InitTarget() start
13:38:15.243;INFO;pylink.jlink;InitTarget() end
13:38:15.443;INFO;pylink.jlink;Found SW-DP with ID 0x2BA01477
13:38:18.568;INFO;pylink.jlink;Unknown DP version. Assuming DPv0
13:38:18.569;INFO;pylink.jlink;Scanning AP map to find all available APs
13:38:19.745;INFO;pylink.jlink;AP[1]: Stopped AP scan as end of AP map has been reached
13:38:19.746;INFO;pylink.jlink;AP[0]: AHB-AP (IDR: 0x24770011)
13:38:19.747;INFO;pylink.jlink;Iterating through AP map to find AHB-AP to use
13:38:21.698;INFO;pylink.jlink;AP[0]: Core found
13:38:21.699;INFO;pylink.jlink;AP[0]: AHB-AP ROM base: 0xE00FF000
13:38:22.724;INFO;pylink.jlink;CPUID register: 0x410FC241. Implementer code: 0x41 (ARM)
13:38:22.724;INFO;pylink.jlink;Found Cortex-M4 r0p1, Little endian.
13:38:23.220;INFO;pylink.jlink;FPUnit: 6 code (BP) slots and 2 literal slots
13:38:23.915;INFO;pylink.jlink;CoreSight components:
13:38:23.915;INFO;pylink.jlink;ROMTbl[0] @ E00FF000
13:38:24.120;INFO;pylink.jlink;ROMTbl[0][0]: E000E000, CID: B105E00D, PID: 000BB00C SCS-M7
13:38:24.223;INFO;pylink.jlink;ROMTbl[0][1]: E0001000, CID: B105E00D, PID: 003BB002 DWT
13:38:24.323;INFO;pylink.jlink;ROMTbl[0][2]: E0002000, CID: B105E00D, PID: 002BB003 FPB
13:38:24.425;INFO;pylink.jlink;ROMTbl[0][3]: E0000000, CID: B105E00D, PID: 003BB001 ITM
13:38:24.526;INFO;pylink.jlink;ROMTbl[0][4]: E0040000, CID: B105900D, PID: 003BB923 TPIU-Lite
13:38:24.625;INFO;pylink.jlink;ROMTbl[0][5]: E0041000, CID: B105900D, PID: 000BB925 ETM
13:38:24.724;INFO;pylink.jlink;InitTarget() start
13:38:26.595;INFO;pylink.jlink;InitTarget() end
13:38:26.794;INFO;pylink.jlink;Found SW-DP with ID 0x2BA01477
13:38:29.996;INFO;pylink.jlink;Unknown DP version. Assuming DPv0
13:38:29.997;INFO;pylink.jlink;AP map detection skipped. Manually configured AP map found.
13:38:29.997;INFO;pylink.jlink;AP[0]: AHB-AP (IDR: Not set)
13:38:31.951;INFO;pylink.jlink;AP[0]: Core found
13:38:31.952;INFO;pylink.jlink;AP[0]: AHB-AP ROM base: 0xE00FF000
13:38:32.935;INFO;pylink.jlink;CPUID register: 0x410FC241. Implementer code: 0x41 (ARM)
13:38:32.936;INFO;pylink.jlink;Found Cortex-M4 r0p1, Little endian.
13:38:33.329;INFO;pylink.jlink;FPUnit: 6 code (BP) slots and 2 literal slots
13:38:34.018;INFO;pylink.jlink;CoreSight components:
13:38:34.018;INFO;pylink.jlink;ROMTbl[0] @ E00FF000
13:38:34.214;INFO;pylink.jlink;ROMTbl[0][0]: E000E000, CID: B105E00D, PID: 000BB00C SCS-M7
13:38:34.314;INFO;pylink.jlink;ROMTbl[0][1]: E0001000, CID: B105E00D, PID: 003BB002 DWT
13:38:34.412;INFO;pylink.jlink;ROMTbl[0][2]: E0002000, CID: B105E00D, PID: 002BB003 FPB
13:38:34.512;INFO;pylink.jlink;ROMTbl[0][3]: E0000000, CID: B105E00D, PID: 003BB001 ITM
13:38:34.612;INFO;pylink.jlink;ROMTbl[0][4]: E0040000, CID: B105900D, PID: 003BB923 TPIU-Lite
13:38:34.717;INFO;pylink.jlink;ROMTbl[0][5]: E0041000, CID: B105900D, PID: 000BB925 ETM
```

**Comment by chanqueo** [comment]
I added some traces in 
`    @open_required
    def connect(self, chip_name, speed='auto', verbose=False):
        """Connects the J-Link to its target.

        Args:
          self (JLink): the ``JLink`` instance
          chip_name (str): target chip name
          speed (int): connection speed, one of ``{5-12000, 'auto', 'adaptive'}``
          verbose (bool): boolean indicating if connection should be verbose in logging

        Returns:
          ``None``

        Raises:
          JLinkException: if connection fails to establish.
          TypeError: if given speed is invalid
        """
        logging.info("EnableRemarks")
        if verbose:
            self.exec_command('EnableRemarks = 1')

        # Determine which device we are.  This is essential for using methods
        # like 'unlock' or 'lock'.
        logging.info("get_device_index")
        index = self.get_device_index(chip_name)

        logging.info("supported_device")
        self._device = self.supported_device(index)

        # This is weird but is currently the only way to specify what the
        # target is to the J-Link.
        logging.info("exec_command")
        self.exec_command('Device = %s' % chip_name)

        # Need to select target interface speed here, so the J-Link knows what
        # speed to use to establish target communication.
        logging.info("set_speed")
        if speed == 'auto':
            self.set_speed(auto=True)
        elif speed == 'adaptive':
            self.set_speed(adaptive=True)
        else:
            self.set_speed(speed)

        logging.info("JLINKARM_Connect")
        result = self._dll.JLINKARM_Connect()
        if result < 0:
            raise errors.JLinkException(result)

        try:
            # Issue a no-op command after connect. This has to be in a try-catch.
            self.halted()
        except errors.JLinkException:
            pass

        return None`

And by calling connect() function I get

`11:06:28.096 INFO     root                 EnableRemarks
11:06:28.096 INFO     root                 get_device_index
11:06:28.106 INFO     root                 supported_device
11:06:28.107 INFO     root                 exec_command
11:06:28.107 INFO     pylink.jlink         Device "EFR32MG12PXXXF1024" selected.
11:06:28.110 INFO     root                 set_speed
11:06:28.113 INFO     pylink.jlink         InitTarget() start
11:06:28.168 INFO     pylink.jlink         InitTarget() end
11:06:28.175 INFO     pylink.jlink         Found SW-DP with ID 0x2BA01477
11:06:28.279 INFO     pylink.jlink         DPv0 detected
11:06:28.279 INFO     pylink.jlink         Scanning AP map to find all available APs
11:06:28.314 INFO     pylink.jlink         AP[1]: Stopped AP scan as end of AP map has been reached
11:06:28.315 INFO     pylink.jlink         AP[0]: AHB-AP (IDR: 0x24770011)
11:06:28.315 INFO     pylink.jlink         Iterating through AP map to find AHB-AP to use
11:06:28.373 INFO     pylink.jlink         AP[0]: Core found
11:06:28.374 INFO     pylink.jlink         AP[0]: AHB-AP ROM base: 0xE00FF000
11:06:28.401 INFO     pylink.jlink         CPUID register: 0x410FC241. Implementer code: 0x41 (ARM)
11:06:28.402 INFO     pylink.jlink         Found Cortex-M4 r0p1, Little endian.
11:06:28.523 INFO     pylink.jlink         FPUnit: 6 code (BP) slots and 2 literal slots
11:06:28.557 INFO     pylink.jlink         CoreSight components:
11:06:28.559 INFO     pylink.jlink         ROMTbl[0] @ E00FF000
11:06:28.574 INFO     pylink.jlink         [0][0]: E000E000 CID B105E00D PID 000BB00C SCS-M7
11:06:28.580 INFO     pylink.jlink         [0][1]: E0001000 CID B105E00D PID 003BB002 DWT
11:06:28.587 INFO     pylink.jlink         [0][2]: E0002000 CID B105E00D PID 002BB003 FPB
11:06:28.594 INFO     pylink.jlink         [0][3]: E0000000 CID B105E00D PID 003BB001 ITM
11:06:28.601 INFO     pylink.jlink         [0][4]: E0040000 CID B105900D PID 003BB923 TPIU-Lite
11:06:28.610 INFO     pylink.jlink         [0][5]: E0041000 CID B105900D PID 000BB925 ETM
11:06:28.610 INFO     root                 JLINKARM_Connect
11:06:28.613 INFO     pylink.jlink         InitTarget() start
11:06:28.666 INFO     pylink.jlink         InitTarget() end
11:06:28.672 INFO     pylink.jlink         Found SW-DP with ID 0x2BA01477
11:06:28.776 INFO     pylink.jlink         DPv0 detected
11:06:28.776 INFO     pylink.jlink         AP map detection skipped. Manually configured AP map found.
11:06:28.777 INFO     pylink.jlink         AP[0]: AHB-AP (IDR: Not set)
11:06:28.831 INFO     pylink.jlink         AP[0]: Core found
11:06:28.831 INFO     pylink.jlink         AP[0]: AHB-AP ROM base: 0xE00FF000
11:06:28.862 INFO     pylink.jlink         CPUID register: 0x410FC241. Implementer code: 0x41 (ARM)
11:06:28.862 INFO     pylink.jlink         Found Cortex-M4 r0p1, Little endian.
11:06:28.971 INFO     pylink.jlink         FPUnit: 6 code (BP) slots and 2 literal slots
11:06:28.991 INFO     pylink.jlink         CoreSight components:
11:06:28.991 INFO     pylink.jlink         ROMTbl[0] @ E00FF000
11:06:28.998 INFO     pylink.jlink         [0][0]: E000E000 CID B105E00D PID 000BB00C SCS-M7
11:06:29.001 INFO     pylink.jlink         [0][1]: E0001000 CID B105E00D PID 003BB002 DWT
11:06:29.006 INFO     pylink.jlink         [0][2]: E0002000 CID B105E00D PID 002BB003 FPB
11:06:29.009 INFO     pylink.jlink         [0][3]: E0000000 CID B105E00D PID 003BB001 ITM
11:06:29.012 INFO     pylink.jlink         [0][4]: E0040000 CID B105900D PID 003BB923 TPIU-Lite
11:06:29.016 INFO     pylink.jlink         [0][5]: E0041000 CID B105900D PID 000BB925 ETM`

**Comment by chanqueo** [comment]
Any progress?

**Comment by hkpeprah** [comment]
Sorry, this sort of got away from. I'm suspecting that maybe the `Device =` is trigger the connect, as I don't see why the `set_speed` would do so. I will try locally to see if I can trace which command, and hopefully get back within the week on a resolution here.

**Comment by hkpeprah** [comment]
I looked a bit at this today, and have traced the code to this section:
```
    # This is weird but is currently the only way to specify what the
    # target is to the J-Link.
    self.exec_command('Device = %s' % chip_name)

    # Need to select target interface speed here, so the J-Link knows what
    # speed to use to establish target communication.
    if speed == 'auto':
        self.set_speed(auto=True)
    elif speed == 'adaptive':
        self.set_speed(adaptive=True)
    else:
        self.set_speed(speed)

    result = self._dll.JLINKARM_Connect()
    if result < 0:
        raise errors.JLinkException(result)
```
Removing the latter half seems to resolve it, but that points to the device assignment doing an auto-connect. Will keep looking into it:
```
    # This is weird but is currently the only way to specify what the
    # target is to the J-Link.
    self.exec_command('Device = %s' % chip_name)

    # Need to select target interface speed here, so the J-Link knows what
    # speed to use to establish target communication.
    if speed == 'auto':
        self.set_speed(auto=True)
    elif speed == 'adaptive':
        self.set_speed(adaptive=True)
    else:
        self.set_speed(speed)

-    result = self._dll.JLINKARM_Connect()
-    if result < 0:
-        raise errors.JLinkException(result)
```

**Comment by hkpeprah** [comment]
Think I have a path forward for this. Will put up a patch later today.

**Comment by hkpeprah** [comment]
Should be fixed in `0.13.0`: https://github.com/square/pylink/releases/tag/v0.13.0
